Data/Vinyl/Index.hs

```diff
@@ -0,0 +1,111 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+
+module Data.Vinyl.Index where
+
+import Data.Vinyl.Core
+import Data.Vinyl.Functor
+
+-- | A Peano index provides a constructive proof that the type-level list @rs@
+-- contains the type @r@.
+data RIndex (rs :: [k]) (r :: k) where
+  Z :: RIndex (r ': rs) r
```

Comment:
Well, I use `Nat` in my projects, so if you remove it, I'll have to redefine it locally and there'll be conflicts. I could, of course, use different names myself, but there are many structures that have zero-ish and succ-ish constructors (to name a few: natural numbers, singleton natural numbers, finite sets, list indices), and if one has to choose which type gets the nice `S` and `Z` constructor names, I say it should be `Nat`.